File: core/src/main/resources/org/treblereel/gwt/three4g/renderers/webgl/WebGLInfo.js
class WebGLInfo {

	/**
	 *  @param {WebGLRenderingContext|WebGL2RenderingContext} gl
	 * 
	 */
	constructor( gl ) {

		this.gl = gl;

		const memory = {
			geometries: 0,
			textures: 0
		};

		const render = {
			frame: 0,
			calls: 0,
			triangles: 0,
			points: 0,
			lines: 0
		};

		this.memory = memory;
		this.render = render;

		this.programs = null;
		this.autoReset = true;

	}

	/**
	 * @param {number} count 
	 * @param {number} mode 
	 * @param {number} instanceCount 
	 */
	update( count, mode, instanceCount ) {

		this.render.calls ++;

		switch ( mode ) {

			case this.gl.TRIANGLES:
				this.render.triangles += instanceCount * ( count / 3 );
				break;

			case this.gl.LINES:
				this.render.lines += instanceCount * ( count / 2 );
				break;

			case this.gl.LINE_STRIP:
				this.render.lines += instanceCount * ( count - 1 );
				break;

			case this.gl.LINE_LOOP:
				this.render.lines += instanceCount * count;
				break;

			case this.gl.POINTS:
				this.render.points += instanceCount * count;
				break;

			default:
				console.error( 'THREE.WebGLInfo: Unknown draw mode:', mode );
				break;

		}

	}

	reset() {

		this.render.frame ++;
		this.render.calls = 0;
		this.render.triangles = 0;
		this.render.points = 0;
		this.render.lines = 0;

	}

}


export { WebGLInfo };

File: core/src/main/java/org/treblereel/gwt/three4g/renderers/webgl/WebGLInfo.java
package org.treblereel.gwt.three4g.renderers.webgl;

import elemental2.core.JsArray;
import elemental2.webgl.WebGLRenderingContext;
import jsinterop.annotations.JsOverlay;
import jsinterop.annotations.JsPackage;
import jsinterop.annotations.JsProperty;
import jsinterop.annotations.JsType;
import jsinterop.base.Js;
import jsinterop.base.JsPropertyMap;

@JsType(isNative = true, name = "THREE.WebGLInfo", namespace = JsPackage.GLOBAL)
public class WebGLInfo {
  @JsType(isNative = true, name = "?", namespace = JsPackage.GLOBAL)
  public interface MemoryFieldType {
    @JsOverlay
    static WebGLInfo.MemoryFieldType create() {
      return Js.uncheckedCast(JsPropertyMap.of());
    }

    @JsProperty
    double getGeometries();

    @JsProperty
    double getTextures();

    @JsProperty
    void setGeometries(double geometries);

    @JsProperty
    void setTextures(double textures);
  }

  @JsType(isNative = true, name = "?", namespace = JsPackage.GLOBAL)
  public interface RenderFieldType {
    @JsOverlay
    static WebGLInfo.RenderFieldType create() {
      return Js.uncheckedCast(JsPropertyMap.of());
    }

    @JsProperty
    double getCalls();

    @JsProperty
    double getFrame();

    @JsProperty
    double getLines();

    @JsProperty
    double getPoints();

    @JsProperty
    double getTriangles();

    @JsProperty
    void setCalls(double calls);

    @JsProperty
    void setFrame(double frame);

    @JsProperty
    void setLines(double lines);

    @JsProperty
    void setPoints(double points);

    @JsProperty
    void setTriangles(double triangles);
  }

  public boolean autoReset;
  public WebGLInfo.MemoryFieldType memory;
  public JsArray<WebGLProgram> programs;
  public WebGLInfo.RenderFieldType render;

  public WebGLInfo(WebGLRenderingContext gl) {}

  public native void reset();

  public native void update(double count, double mode, double instanceCount);
}
